tools/cat2table.ml

```diff
@@ -0,0 +1,883 @@
+  | Op1 _, _ -> -1
+  | _, Op1 _ -> 1
+  | Op (_, op1, es1), Op (_, op2, es2) ->
+    let c = compare op1 op2 in
+    if c != 0 then c
+    else compare_expr_list es1 es2
+  | Op _, _ -> -1
+  | _, Op _ -> 1
+  | _, _ -> 0
+
+  and compare_expr_list es1 es2 = match es1, es2 with
+  | [], [] -> 0
+  | [], _ -> -1
+  | _, [] -> 1
+  | e1 :: e1s, e2 :: e2s ->
+    let c = compare_expr e1 e2 in
+    if c != 0 then c
+    else compare_expr_list e1s e2s
+
+  let rec sort_subexprs = function
+  | (Konst (_) | Var (_)) as e -> e
+  | Op1 (loc, op, e) -> Op1 (loc, op, sort_subexprs e)
+  | Op (loc, op, es) ->
+    let es = List.map sort_subexprs es in
+    let es = List.sort compare_expr es in
+    Op (loc, op, es)
+  | e -> e
+
+  let expr_equal e1 e2 =
+    let do_op op1 op2 e1 e2 =
+      if op1 != op2 then false
+      else
+        let e1 = sort_subexprs e1 in
+        let e2 = sort_subexprs e2 in
+        compare_expr e1 e2 = 0 in
+    match e1, e2 with
+    | e1, e2 when is_empty e1 && is_empty e2 -> true
+    | Konst (_, (Universe _)), Konst (_, (Universe _)) -> true
+    | Var (_, id1), Var (_, id2) -> id1 = id2
+    | Op (_, op1, _), Op (_, op2, _) -> do_op op1 op2 e1 e2
+    | Op1 (_, op1, e1), Op1 (_, op2, e2) -> do_op op1 op2 e1 e2
+    | _, _ -> false
+
+  let barriers = ["DMB.ISH"; "DMB.ISHLD"; "DMB.ISHST"; "DMB.LD"; "DMB.OSH";
+    "DMB.OSHLD"; "DMB.OSHST"; "DMB.ST"; "DMB.SY"; "DSB.ISH"; "DSB.ISHLD";
+    "DSB.ISHST"; "DSB.LD"; "DSB.OSH"; "DSB.OSHST"; "DSB.ST"; "DSB.SY"; "ISB"]
+
+  let pte_accesses = ["PTEDevice"; "PTEDevice-GRE"; "PTEDevice-nGRE";
+    "PTEDevice-nGnRE"; "PTEDevice-nGnRnE"; "PTEInner-non-cacheable";
+    "PTEInner-shareable"; "PTEInner-write-back"; "PTEInner-write-through";
+    "PTENon-shareable"; "PTENormal"; "PTEOuter-non-cacheable";
+    "PTEOuter-shareable"; "PTEOuter-write-back"; "PTEOuter-write-through";
+    "PTETaggedNormal"; "PTEXS"]
+
+  let pte_vals = ["PTEINV"; "PTEV"]
+  let all_ptes = "PTEMemAttr" :: (pte_vals @ pte_accesses)
+
+  let supersets = [
+    ("A", ["M"; "R"; "Exp"]);
+    ("AF", ["NExp"]);
+    ("AccessFlag", ["MMU"; "EXC-ENTRY"; "FAULT"]);
+    ("BCC", ["B"]);
+    ("DB", ["NExp"]);
+    ("EXC-ENTRY", ["FAULT"]);
+    ("EXC-RET", ["B"]);
+    ("Instr", ["M"]);
+    ("IW", ["W"; "M"]);
+    ("L", ["M"; "W"; "Exp"]);
+    ("MMU", ["EXC-ENTRY"; "FAULT"]);
+    ("NoRet", ["R"; "M"]);
+    ("PA", ["M"]);
+    ("Permission", ["MMU"; "EXC-ENTRY"; "FAULT"]);
+    ("PTE", ["M"]);
+    ("PTEMemAttr", ["PTE"; "M"]);
+    ("Q", ["M"; "R"; "Exp"]);
+    ("R", ["M"]);
+    ("Restricted-CMODX", ["Instr"; "M"]);
+    ("SupervisorCall", ["EXC-ENTRY"; "FAULT"]);
+    ("T", ["M"]);
+    ("TagCheck", ["FAULT"]);
+    ("Translation", ["MMU"; "EXC-ENTRY"; "FAULT"]);
+    ("TLBIIS", ["TLBI"]);
+    ("TLBInXS", ["TLBI"]);
+    ("UndefinedInstruction", ["EXC-ENTRY"; "FAULT"]);
+    ("W", ["M"]);
+  ] @ List.map (fun b -> b, ["F"]) barriers
+    @ List.map (fun e -> e, ["PTE"; "M"; "PTEMemAttr"]) pte_accesses
```

Comment:
I think I understand why you need this but it would be really good if there was a way to get it programmatically rather than having it hardcoded.

Reply:
I understand, we might have to do some refactoring elsewhere, but as it stands, it's very difficult to ensure this correct/complete and it will be impossible to keep it in sync with changes in herd7.